addons/binding/org.openhab.binding.mihome.test/src/test/groovy/org/openhab/binding/mihome/discovery/test/MiHomeSubdeviceDiscoveryOSGiTest.groovy

```diff
@@ -0,0 +1,210 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.mihome.discovery.test
+
+import static org.hamcrest.CoreMatchers.*
+import static org.junit.Assert.*
+
+import javax.servlet.http.HttpServlet
+
+import org.eclipse.smarthome.config.discovery.DiscoveryResult
+import org.eclipse.smarthome.config.discovery.DiscoveryService
+import org.eclipse.smarthome.config.discovery.inbox.Inbox
+import org.eclipse.smarthome.core.thing.Bridge
+import org.eclipse.smarthome.core.thing.Thing
+import org.eclipse.smarthome.core.thing.ThingRegistry
+import org.eclipse.smarthome.core.thing.ThingStatus
+import org.eclipse.smarthome.core.thing.ThingTypeUID
+import org.eclipse.smarthome.core.thing.ThingUID
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.openhab.binding.mihome.MiHomeBindingConstants
+import org.openhab.binding.mihome.handler.MiHomeGatewayHandler
+import org.openhab.binding.mihome.handler.MiHomeSubdevicesHandler
+import org.openhab.binding.mihome.internal.api.JSONResponseHandler
+import org.openhab.binding.mihome.internal.api.constants.DeviceTypesConstants
+import org.openhab.binding.mihome.internal.api.manager.*
+import org.openhab.binding.mihome.internal.discovery.MiHomeSubdeviceDiscoveryService
+import org.openhab.binding.mihome.test.AbstractMiHomeOSGiTest
+import org.openhab.binding.mihome.test.JsonGateway
+import org.openhab.binding.mihome.test.JsonSubdevice
+import org.openhab.binding.mihome.test.MiHomeServlet
+
+import com.google.gson.JsonObject
+/**
+ * Tests for the {@link MiHomeSubdeviceDiscoveryService}
+ *
+ * @author Svilen Valkanov
+ *
+ */
+class MiHomeSubdeviceDiscoveryTest extends AbstractMiHomeOSGiTest{
+
+    private Thing subdeviceThing
+    private MiHomeSubdeviceDiscoveryService discoveryService;
+    private Inbox inbox
+    private JsonObject serverResponseBody
+
+    @Before
+    public void setUp() {
+        setUpServices()
+
+        // in order to test subdevices we need to register servlet representing successful gateway registration
+        JsonGateway gatewayDevice = new JsonGateway()
+        String content = generateShowJsonDeviceServerResponse("success", gatewayDevice)
+        MiHomeServlet successfullGatewayRegistrationServlet = new MiHomeServlet(content)
+        registerServlet(PATH_CREATE_GATEWAY, successfullGatewayRegistrationServlet)
+
+        gatewayThing = createBridge(thingRegistry,TEST_PASSWORD,TEST_USERNAME,TEST_GATEWAY_CODE,TEST_GATEWAY_ID)
+        assertThat gatewayThing, is(notNullValue())
```

Comment:
Only few lines below we have a waitForAssert that is necessary. 

I am using it to verify that the framework has done something (initialized the Thing) before continuing with setting the mocked client.